### PR TITLE
[Utopia-562] add filter pia by drafter/status/ministry in backend

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

-add backend function for filter
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [x] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [x] N/A
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

## Definition of Done

Definition of Done

### CODE REVIEW

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts [hunk 1]
```diff
@@ -25,6 +27,37 @@ export class PiaIntakeFindQuery {
...
+
+  @ApiProperty({
+    required: false,
+    type: String,
+    example: 'onlyMyPia',
+  })
+  @IsString()
+  @IsOptional()
+  @Length(0, 20)
+  readonly filterByDrafter?: string = '';
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:59]: why default to empty string?

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts [hunk 2]
```diff
@@ -25,6 +27,37 @@ export class PiaIntakeFindQuery {
...
 
+  @ApiProperty({
+    required: false,
+    type: PiaIntakeStatusEnum,
+    example: piaIntakeEntityMock.status,
+  })
+  @IsString()
+  @IsOptional()
+  @Length(0, 20)
+  @Transform(({ value }) => value.trim())
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:38]: remove transform and length

add IsIn(PiaIntakeStatusEnum)

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts [hunk 3]
```diff
@@ -25,6 +27,37 @@ export class PiaIntakeFindQuery {
...
+  readonly filterByStatus?: string = '';
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:39]: no need for default

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts [hunk 4]
```diff
@@ -25,6 +27,37 @@ export class PiaIntakeFindQuery {
...
+
+  @ApiProperty({
+    required: false,
+    type: GovMinistriesEnum,
+    example: piaIntakeEntityMock.ministry,
+  })
+  @IsString()
+  @IsOptional()
+  @Length(0, 100)
+  readonly filterByMinistry?: string = '';
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:49]: remove length and default empty string

add IsIn(GovMinistriesEnum)

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts (hunk 1 above)
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:59]: filterByDrafter isn't a right naming terminology. Also, what is the data that it expects? Please create an enum of it, allowing any random string.

File: src/backend/src/modules/pia-intake/pia-intake.service.ts [hunk 5]
```diff
@@ -128,6 +129,7 @@ export class PiaIntakeService {
     const whereClause: FindOptionsWhere<PiaIntakeEntity>[] = [];
 
     // Scenario 1: As a user, retrieve all PIA-intakes I submitted
+
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/pia-intake.service.ts:132]: unnecessary line break

File: src/backend/src/modules/pia-intake/pia-intake.service.ts [hunk 6]
```diff
@@ -162,6 +163,30 @@ export class PiaIntakeService {
 
       whereClause.push(...additionalWhereClauses);
     }
+
+    /** filter logic here */
+    if (query.filterByStatus) {
+      whereClause.forEach((clause) => {
+        clause.status = PiaIntakeStatusEnum[query.filterByStatus];
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/pia-intake.service.ts:170]: shouldn't this be `clause.status = query.filterByStatus`?

File: src/backend/src/modules/pia-intake/pia-intake.service.ts [hunk 7]
```diff
@@ -162,6 +163,30 @@ export class PiaIntakeService {
...
+      });
+    }
+    if (query.filterByMinistry) {
+      whereClause.forEach((clause) => {
+        clause.ministry = GovMinistriesEnum[query.filterByMinistry];
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/pia-intake.service.ts:175]: similarly here

File: src/backend/src/modules/pia-intake/pia-intake.service.ts [hunk 8]
```diff
@@ -162,6 +163,30 @@ export class PiaIntakeService {
...
+      whereClause.forEach((clause) => {
+        clause.createdByGuid = Not(user.idir_user_guid);
+      });
+    }
+    // filter by drafter sub scenario 2 check the filter to get only my Pia
+    if (query.filterByDrafter && query.filterByDrafter === 'onlyMyPia') {
+      whereClause.forEach((clause) => {
+        clause.createdByGuid = user.idir_user_guid;
+      });
+    }
```
Comment (kushal-arora-fw) [inline src/backend/src/modules/pia-intake/pia-intake.service.ts:189]: can you please create a meaningful enum for excludeMyPia, onlyMyPia?

**Review by kushal-arora-fw — COMMENTED** [review]
Add some improvement comments. + it is missing unit test cases

File: src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts (hunk 1 above)
Comment (binggaofw) [reply src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:59]: removed all the default value

Comment (binggaofw) [reply src/backend/src/modules/pia-intake/dto/pia-intake-find-query.dto.ts:59]: add a new enum PiaFilterDrafterByCurrentUserEnum, change to filterPiaDrafterByCurrentUser

File: src/backend/src/modules/pia-intake/pia-intake.service.ts (hunk 6 above)
Comment (binggaofw) [reply src/backend/src/modules/pia-intake/pia-intake.service.ts:170]: it throw an error

Screen Shot 2023-01-24 at 4 08 38 PM

**Review by kushal-arora-fw — APPROVED** [review]

**Review by Bottle7 — APPROVED** [review]

**Pull request merged**